### PR TITLE
Initial P4 code to support network policies

### PR DESCRIPTION
**Pull request by dhruvpt**
Signed-off-by: Tripathi, Dhruv Prakash <dhruv.tripathi@intel.com>

### CODE REVIEW

File: k8s_dp/k8s_dp.p4 [hunk 1]
```diff
@@ -501,69 +575,464 @@ control k8s_dp_control(
...
+            meta.acl_pol_id = pol_id;
+        }
+
+        action set_status_deny_all() {
+            P_cntr.count(ACL_BLOCKED);
+            acl_deny();
+        }
+
+        table acl_srcip_proto_table {
+            key = {
```
Comment (shweta503) [inline k8s_dp/k8s_dp.p4:603]: We decided to have two SEM tables - acl_srcip_proto_table and acl_srcip_table. The latter would match only on the src ip address to take care of the scenario where only ipset is specified in the rule and not protocol (and port range). The "ipset only" scenario is not handled by having just this table.

File: k8s_dp/k8s_dp.p4 [hunk 2]
```diff
@@ -501,69 +575,464 @@ control k8s_dp_control(
...
+            actions = {
+                set_status_lookup_ipset_only;
+                set_range_check_ref;
+                set_status_deny_all;
+                NoAction;   /* set_status_allow_all */
+            }
+            const default_action = NoAction();
+        }
+
+        table acl_dstip_proto_table {
```
Comment (shweta503) [inline k8s_dp/k8s_dp.p4:616]: Same as above comment

File: k8s_dp/k8s_dp.p4 [hunk 3]
```diff
@@ -501,69 +575,464 @@ control k8s_dp_control(
...
-		}	
+                /* Assumption: Intra-cluster traffic from one pod to another.
+                 * For initial HOST_TO_NET direction, apply egress network policy.
+                 * For service to client, this is before SNAT w.r.t. endpoint pod.
+                 * For client to service, this is before DNAT w.r.t. client pod
+                 * with service IP as remote IP.
+                 */
+                if ((hdr.ipv4.isValid()) &&
+                    (meta.direction == PNA_Direction_t.HOST_TO_NET))
+                {
```
Comment (shweta503) [inline k8s_dp/k8s_dp.p4:841]: The order of applying DNAT and egress policy seems wrong here. In TX direction (HOST_TO_NET), we apply egress policy *after* load balancing/DNAT.

File: k8s_dp/k8s_dp.p4 [hunk 4]
```diff
@@ -501,69 +575,464 @@ control k8s_dp_control(
...
+                 * Assumption: Intra-cluster traffic from one pod to another.
+                 * After port loopback, in NET_TO_HOST direction, apply ingress
+                 * network policy w.r.t. dst pod.
+                 * For client to service, this is after DNAT w.r.t. endpoint pod.
+                 * For service to client, this is after SNAT w.r.t. client pod
+                 * with service IP as remote IP.
+                 */
+                if ((hdr.ipv4.isValid()) &&
+                    (meta.direction == PNA_Direction_t.NET_TO_HOST))
+                {
```
Comment (shweta503) [inline k8s_dp/k8s_dp.p4:985]: Again the order seems wrong. I believe, the correct order of the apply block should be:
1. If (RX), apply ingress policy
2. Load balance/DNAT or SNAT
3. If (TX), apply egress policy

**Review by shweta503 — CHANGES_REQUESTED** [review]

File: k8s_dp/k8s_dp.p4 (hunk 1 above)
Comment (dhruvpt) [reply k8s_dp/k8s_dp.p4:603]: In both these tables, the second protocol field is LPM and thus, can be completely wildcarded to have L4 protocol independent policy. So, same response for below comment too

File: k8s_dp/k8s_dp.p4 (hunk 3 above)
Comment (dhruvpt) [reply k8s_dp/k8s_dp.p4:841]: The egress policy is w.r.t. to source/originating pod and destination network. The destination network can be cluster IP address to filter access to a service without having to add individual rules for all end points. That requires filtering before DNAT. This is esp. helpful when end points can change dynamically. Let's talk further on this.

File: k8s_dp/k8s_dp.p4 (hunk 4 above)
Comment (dhruvpt) [reply k8s_dp/k8s_dp.p4:985]: We are here assuming only pod-to-pod traffic at present. So, any processing begins with egress filter for originating pod (which can be client or a service end-point). Thereafter, after DNAT (or SNAT), packet is looped back to re-enter as ingress traffic. At this point, all packet modifications have already been done and only thing remaining is the ingress filtering w.r.t. destination pod (which can be client pod or a service end-point). The remote IP, for filtering purpose, can now be cluster IP as well. Again, let's talk further on this.

File: k8s_dp/k8s_dp.p4 (hunk 3 above)
Comment (shweta503) [reply k8s_dp/k8s_dp.p4:841]: I see the change from exact to lpm. thanks.

File: k8s_dp/k8s_dp.p4 (hunk 4 above)
Comment (shweta503) [reply k8s_dp/k8s_dp.p4:985]: Yes, with change in design from applying policy post-DNAT to pre-DNAT, this makes sense.

**Review by shweta503 — APPROVED** [review]

File: k8s_dp/k8s_dp.p4 [hunk 5]
```diff
@@ -26,6 +46,16 @@ const ActionRef_t WRITE_SRC_IP = (ActionRef_t) 1;
...
 const ActionRef_t NO_MODIFY =  (ActionRef_t) 0;
 
+typedef bit<8> RuleMatchMask_t;
+typedef bit<8> RangeCheckRefType_t;
+typedef bit<8> AclPolicyId_t;
+
+const RangeCheckRefType_t CHECK_TCP_DST_PORT_RANGE = (RangeCheckRefType_t) 1;
+const RangeCheckRefType_t CHECK_UDP_DST_PORT_RANGE = (RangeCheckRefType_t) 2;
+const RangeCheckRefType_t CHECK_SCTP_DST_PORT_RANGE = (RangeCheckRefType_t) 3;
+const RangeCheckRefType_t CHECK_UDPL_DST_PORT_RANGE = (RangeCheckRefType_t) 4;
```
Comment (nupurjai) [inline k8s_dp/k8s_dp.p4:56]: For UDPL, we need checksum checks. Parser is not doing it right now.

File: k8s_dp/k8s_dp.p4 [hunk 6]
```diff
@@ -501,69 +575,464 @@ control k8s_dp_control(
 		const default_action = NoAction();
 	}
 	
+        action acl_deny() {
+            P_cntr.count(ACL_DENIED);
```
Comment (nupurjai) [inline k8s_dp/k8s_dp.p4]: keep it lowercase.

File: k8s_dp/k8s_dp.p4 (hunk 5 above)
Comment (dhruvpt) [reply k8s_dp/k8s_dp.p4:56]: I assume that's a general comment as in this commit for policy, we are not doing anything related to checksum of any type.

File: k8s_dp/k8s_dp.p4 (hunk 6 above)
Comment (dhruvpt) [reply k8s_dp/k8s_dp.p4]: I assume, you are referring to P_cntr

File: README.md [hunk 7]
```diff
@@ -381,12 +379,14 @@ Kubernetes is known to not work well with Linux swap and hence, it should be tur
...
-  ip link set P4TAP_0 name TAP_0
-  ip link set TAP_0 up
-  ip addr add 169.254.1.1/32 dev TAP_0
-  ARP_PROXY_IF=TAP_0 ./bin/arp_proxy &
+  ip netns add arpns
+  ip link set P4TAP_0 netns arpns
+  ip netns exec arpns ip link set P4TAP_0 up
+  ip netns exec arpns ip addr add 169.254.1.1/32 dev P4TAP_0
+  ip netns exec arpns ip route add default dev P4TAP_0 scope link
+  ip netns exec arpns ARP_PROXY_IF=P4TAP_0 ./bin/arp_proxy &
```
Comment (jokuniew) [inline README.md:389]: This line keeps telling me that `exec of "ARP_PROXY_IF=P4TAP_0" failed: No such file or directory`
When I do `ip netns exec arpns ls ./bin` I can see the arp_proxy binary

It used to work on previous version. Shouldn't be there any more changes done?

Comment (nupurjai) [reply README.md:389]: ip netns exec arpns export ARP_PROXY_IF=P4TAP_0 && ./bin/arp_proxy &

**Pull request closed without merging**